File: 2022/day_01/puzzle.js
const { data_test } = require('./data_test.js');
const { puzzle_data } = require('./puzzle_data.js');

function compute_data(data) {
    var res = []
    var elfBag = []

    var split = data.split("\n");
    for (var i=0; i<split.length; i++) {
        let s = split[i]
        if (s === "") {
            res.push(elfBag)
            elfBag = []
        } else {
            elfBag.push(s)
        }
    }
    res.push(elfBag)
    return res;
}

function part1(data) {
    var sumBag = []
    for (var i=0; i<data.length; i++) {
        let bag = data[i]

        var sum = 0
        for (var j=0; j<bag.length; j++) {
            sum += Number(bag[j])
        }
        sumBag.push(sum)
    }
    return Math.max(...sumBag)
}

function part2(data) {
    var sumBag = []
    for (var i=0; i<data.length; i++) {
        let bag = data[i]

        var sum = 0
        for (var j=0; j<bag.length; j++) {
            sum += Number(bag[j])
        }
        sumBag.push(sum)
    }
    sumBag.sort(function(a,b) { return a - b; })

    return sumBag.slice(-3).reduce((partialSum, a) => partialSum + a, 0)
}


let data = compute_data(puzzle_data)
console.log('Result for puzzle 1 = ', part1(data))
console.log('Result for puzzle 2 = ', part2(data))

File: 2024/day_05/puzzle.py
def read_data(filename):
    return open(f"{filename}").read().strip()

###############################
## Construct data for solver ##
###############################
def constructData(data):
    rule_lists, page_lists = data.split("\n\n")
    rules = [ list(r.split('|')) for r in rule_lists.splitlines() ]
    rules = [ list(map(int, r)) for r in rules ]
    updates = [ list(r.split(',')) for r in page_lists.splitlines() ]
    updates = [ list(map(int, p)) for p in updates ]

    return [rules, updates]


###########
## UTILS ##
###########
def is_ok(rules, listOfPages):
    for rule in rules:
        if not set(rule).issubset(set(listOfPages)): 
            continue
            
        left, right = rule
        left_i = listOfPages.index(left)
        right_i = listOfPages.index(right)
        
        if left_i > right_i:
            return False
    
    return True


def sorted_list_of_page(rules, listOfPages):
    listOrdered = listOfPages.copy()
    
    is_update = True
    while is_update:
        is_update = False
        for rule in rules:
            if not set(rule).issubset(set(listOrdered)): 
                continue
                
            left, right = rule
            left_i = listOrdered.index(left)
            right_i = listOrdered.index(right)
            
            if left_i > right_i:
                listOrdered[left_i], listOrdered[right_i] = listOrdered[right_i], listOrdered[left_i]
                is_update = True
    
    return listOrdered


###################
## Solver part 1 ##
###################
def part1(data):
    rules = data[0]
    listToUpdate = data[1]
    
    res = []
    for l in listToUpdate:
        if is_ok(rules, l):
            mid = l[len(l)//2]
            res.append(mid)
    
    return sum(res)


###################
## Solver part 2 ##
###################
def part2(data):
    rules = data[0]
    listToUpdate = data[1]
    
    unorder = []
    for l in listToUpdate:
        if not is_ok(rules, l):
            unorder.append(l)
    
    res = []
    for l in unorder:
        mid = sorted_list_of_page(rules, l)[len(l)//2]
        res.append(mid)
    return sum(res)


##########
## MAIN ##
##########
data = read_data("puzzle_data.txt")
print('Result for puzzle 1 = ', part1(constructData(data)))
print('Result for puzzle 2 = ', part2(constructData(data)))